File: android/edge-frameworks-core/src/main/kotlin/io/github/rajibsarwar/edgeframeworks/EdgeRagModels.kt
package io.github.rajibsarwar.edgeframeworks

data class EdgeDocument(
    val id: String,
    val text: String,
    val metadata: Map<String, String> = emptyMap()
)

data class EdgeChunk(
    val id: String,
    val documentId: String,
    val text: String,
    val metadata: Map<String, String> = emptyMap()
)

data class EdgeEmbedding(
    val values: List<Float>
)

data class EdgeSearchResult(
    val chunk: EdgeChunk,
    val score: Float
)

sealed class EdgeVectorException(message: String) : Exception(message) {
    data class DimensionMismatch(
        val expected: Int,
        val actual: Int
    ) : EdgeVectorException(
        "Embedding dimension mismatch: expected $expected, got $actual"
    )

    data class CountMismatch(
        val expected: Int,
        val actual: Int
    ) : EdgeVectorException(
        "Item count mismatch: expected $expected, got $actual"
    )

    data object ZeroMagnitude :
        EdgeVectorException("Cosine similarity is undefined for a zero-magnitude vector")
}

File: ios/Sources/EdgeFrameworks/Core/EdgeRAGModels.swift
public struct EdgeDocument: Equatable, Sendable {
    public let id: String
    public let text: String
    public let metadata: [String: String]

    public init(
        id: String,
        text: String,
        metadata: [String: String] = [:]
    ) {
        self.id = id
        self.text = text
        self.metadata = metadata
    }
}

public struct EdgeChunk: Equatable, Sendable {
    public let id: String
    public let documentID: String
    public let text: String
    public let metadata: [String: String]

    public init(
        id: String,
        documentID: String,
        text: String,
        metadata: [String: String] = [:]
    ) {
        self.id = id
        self.documentID = documentID
        self.text = text
        self.metadata = metadata
    }
}

public struct EdgeEmbedding: Equatable, Sendable {
    public let values: [Float]

    public init(values: [Float]) {
        self.values = values
    }
}

public struct EdgeSearchResult: Equatable, Sendable {
    public let chunk: EdgeChunk
    public let score: Float

    public init(chunk: EdgeChunk, score: Float) {
        self.chunk = chunk
        self.score = score
    }
}

public enum EdgeVectorError: Error, Equatable, Sendable {
    case dimensionMismatch(expected: Int, actual: Int)
    case countMismatch(expected: Int, actual: Int)
    case zeroMagnitude
}
